proposals/dark_mode_extension_icons.md

```diff
@@ -0,0 +1,227 @@
+  
+  The icon paths are set using the size as key and/or the keyword "any".
+  
+  Optionally the `color_scheme` matching criterium which can either be an string or array of color schemes.
+Set icon_variants dynamically.
+```
+const exampleProperties: {string: string | ImageData}[] = [
+  {
+    "any": "any.svg",
+  },
+  {
+    "16": "16.png",
+    "32": "32.png"
+  },
+  {
+    "16": darkImageData16,
+    "32": darkImageData32,
+    "color_scheme": "dark"
+  },
+  {
+    "16": lightImageData16,
+    "32": lightImageData32,
+    "color_scheme": "light"
+  }
+];
+
+// action.setIcon().
+const createProperties = {variants: exampleProperties};
+action.setIcon(createProperties);
+
+// menus.*(), for supporting browsers.
+const menusProperties = {icon_variants: exampleProperties};
+menus.create(menusProperties);
+menus.update(menusProperties);
+```
+
+A benefit of this new structure is that it's more resiliant to future changes,
+thus allowing for more keys such as density (e.g. 2dppx), purpose (e.g.
+monochrome), and etc.
+
+## setIcon()
+Incumbent action.setIcon(), for reference.
+```
+action.setIcon({
+  path,
+  tabId
+  imageData,
+});
+```
+
+### Behavior
+
+Existing manifest key. The behavior of the icons manifest key will remain
+unchanged.
+```
+{
+  "icons": {
+    "64": "icon_64.png"
+  },
+  "action": {
+    "default_icon": {
+      "64": "action_64.png"
+    }
+  }
+}
+```
+
+Order of precedence. The new `icon_variants` keys and subkeys will take
+precedent, followed by the incumbent `icons` key.
+
+The `dark` icon variant can be used by browsers to improve readability of the
+icon. Often this will be when the OS or browser color scheme is dark (often
+referred to as "dark mode") but also when the browser uses a darker shade theme
+which has a background on which the dark icon variant would be more readable.
+The `light` icon variant is used otherwise.
+
+### New Permissions
+
+N/A.
+
+### Manifest File Changes
+
+1. Any icon group that does not contain a "color_scheme" key will apply to all
+available options, e.g. both "light" and "dark".
+
+1. If "icon_variants" contains an icon group with matching conditions, the icon(s)
+specified in the first matching icon group based on insertion order will be used.
+The other icon groups after that match will be ignored.
+   * "\<size\>" will be used used instead of `"any"` in case in case there are
+   matching conditions.
+   * If there is more than one matching icon object, any without `color_scheme` will be applied to every possible color scheme. Therefore, a subsequent matching icon object with a `color_scheme` will not be used.
+   * Non-exact size matches will return an icon nearest the requested size, or the smallest in case of a tie.
+   * Incorrectly typed `color-scheme` will be ignored, with an optional warning.
+   * If e.g. only one icon object is defined with a specific color scheme, that
+   icon object will be applied to all color schemes. It will be treated as the
+   default.
```

Comment:
```suggestion
   * If only one icon object is defined with a specific color scheme, that
   icon object will be applied to all color schemes. It will be the icon used.
```